File: Zig/Practice/find_multiples.zig
const std = @import("std");
const print = std.debug.print;

pub fn main() !void {
    const array = [_]i32{ 10, 6, 5, 6, 20 };
    const length = array.len;

    const target: i32 = 0;

    if (target == 0) {
        print("Target cannot be 0.\n", .{});
        return;
    }

    const matched = findMultiples(&array, 0, length - 1, target);
    if (matched > 0) {
        print("There are {} multiple(s) of {}.\n", .{ matched, target });
    } else {
        print("There are no multiples of {}.\n", .{target});
    }
}

fn findMultiples(array: []const i32, start: usize, end: usize, target: i32) usize {
    var count: usize = 0;

    if (@mod(array[start], target) == 0) {
        count += 1;
    }

    if (start == end) {
        return count;
    }

    return count + findMultiples(array, start + 1, end, target);
}

File: C/Practice/find_multiples.c
#include <stdio.h>

int find_multiples(const int array[], int start, int end, int target);

int main()
{
    int array[] = {10, 6, 4, 5, 20};
    size_t length = sizeof(array) / sizeof(array[0]);

    int target = 2;

    if (target == 0) {
        printf("Target cannot be zero.\n");
        return 1;
    }

    int matched = find_multiples(array, 0, length - 1, target);
    if (matched > 0)
    {
        printf("There are %d multiple(s) of %d.\n", matched, target);
    }
    else
    {
        printf("There are no multiples of %d.\n", target);
    }

    return 0;
}

int find_multiples(const int array[], int start, int end, int target)
{
    int count = 0;

    if (array[start] % target == 0)
    {
        count++;
    }

    if (start == end)
    {
        return count;
    }

    return count + find_multiples(array, start + 1, end, target);
}